aspnetcore/mvc/advanced/custom-formatters.md

```diff
@@ -1,18 +1,101 @@
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=classdef)]
+
+For binary types, derive from the [InputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.inputformatter) or [OutputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.outputformatter) base class.
+
+### Specify valid media types and encodings
+
+In the constructor, specify valid media types and encodings by adding to the `SupportedMediaTypes` and `SupportedEncodings` collections.
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=ctor&highlight=3,5-6)]
+
+> [!NOTE]  
+> You can't do constructor dependency injection in a formatter class. For example, you can't get a logger by adding a logger parameter to the constructor. To access services, you have to use the context object that gets passed in to your methods. A code example [below](#read-write) shows how to do this.
+
+### Override CanReadType/CanWriteType 
+
+Specify the type you can deserialize into or serialize from by overriding the `CanReadType` or `CanWriteType` methods. For example, you might only be able to create vCard text from a `Contact` type and vice versa.
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=canwritetype)]
+
+For output formatters, in some scenarios you have to override `CanWriteResult` instead of `CanWriteType`. You need to use `CanWriteResult` if your action method returns a model class, and there are derived classes, and you need to know at runtime which derived class was returned by the action.  For example, suppose your action method signature returns a `Person` type, but it may return a `Student` or `Instructor` type that derives from `Person`. If you want your formatter to handle only `Student` objects, check the type of [Object](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.outputformattercanwritecontext#Microsoft_AspNetCore_Mvc_Formatters_OutputFormatterCanWriteContext_Object) in the context object provided to the `CanWriteResult` method. Note that this only applies when the action method returns a model type; if the action method returns `IActionResult`, the `CanWriteType` method receives the runtime type.
+
+<a id="read-write"></a>
+### Override ReadRequestBodyAsync/WriteResponseBodyAsync 
+
+You do the actual work of deserializing or serializing in `ReadRequestBodyAsync` or `WriteResponseBodyAsync`.  The highlighted lines in the following code example show how to get services from the dependency injection container, since you can't get them from constructor parameters.
```

Comment:
Does this make the sentence technically shorter?

The highlighted lines in the following code  show how to get services from the dependency injection container (since you can't get them from constructor parameters).